geode-core/src/main/java/org/apache/geode/cache/query/internal/CompiledIndexOperation.java

```diff
@@ -104,7 +105,13 @@ public Object evaluate(ExecutionContext context) throws TypeMismatchException,
     }
 
     if (rcvr instanceof Map) {
-      return ((Map) rcvr).get(index);
+      if (context instanceof QueryExecutionContext) {
```

Comment:
So when we execute a query and the index is a map index... if the key is not present in the map, we could get an actual "null" back.  When we were adding the NULL key as a collection into the map, we were possibly getting a valid collection?

I'm surprised a bit that the != queries work but I haven't walked through this code in a while.  SInce the tests are passing, that's great.  I would assume the != queries will now be doing a full region crawl... if that's the case, they will be a lot slower.. not sure if that's the case, but if it is, it might not be acceptable => might need to convince the dev list on this one...

Reply:
I have pushed a new commit with some test cases to compare the results of queries with the indexes and without them.
As you can see, the tests when indexes are used fail.
The funny thing is that these test cases also fail if run against the latest code in development. So I think there is something not working right with the indexes, with or without my change.